### ISSUE TITLE
Memory leak on DataSourceDepedentCache

### ISSUE THREAD

**Issue by komdil** [OData/AspNetCoreOData#160]
Unfortunately, I can't provide a source code and I couln't reproduce this problem in simple project. BUT tested it by dotMemory. 
DataSourceDepedentCache is taking about 700mb memory. I think it should be less. Here is snapshot:

![image](https://user-images.githubusercontent.com/50167116/117103599-10969080-ad94-11eb-8c8b-3fecfee9f569.png)
![image](https://user-images.githubusercontent.com/50167116/117103747-52273b80-ad94-11eb-9d49-9f66cfc7a812.png)

**Comment by komdil** [comment]
@xuzhg  I could reproduce this issue. It is happening when we have dynamic prefix and many controllers. To reproduce this issue you can use this project:  https://github.com/komdil/ODataCoreTest/tree/memoryLeak
It has 200 controllers and entities. When we send request it will take more than 7gb memory. Also it is freezing.

**Comment by xuzhg** [comment]
@komdil 
I forward your issue to ASP.NET Core team since the logic of DataSourceDependentCache is not from OData.

Here's the comments:

```txt
Hi,

It's not a memory leak. They are producing a gigantic route table. I'll have to look at this a bit more in detail. 
We've seen other cases like this, where people were registering 40K routes and building these massive route tables. 

There are at a minimum 1400 actions/endpoints on that tree, which is around the size these things can start to get tricky.

There are a few optimizations that we are considering on this area that will help "prune" the tree of unnecessary branches
which can make the tree significantly smaller, we've got other reports of similar issues, however they involved much larger sets of actions.

I'll look in more detail to see why what's exactly at play here and see what we can do to mitigate this in 6.0.
```

another comments:

```txt
it’s only a memory leak if it grows unbounded over time (can’t tell from a single snapshot). 
Otherwise it’s just memory use 😊. Nothing wrong with that.
```

**Comment by komdil** [comment]
@xuzhg  I don't think that it is related to ASP.NET Core. It is happening becasue of dynamic prefix in AddOModel:

```
 opt.AddModel("{contextToken}", edmModel, configureAction =>
                {
                    configureAction.AddService(Microsoft.OData.ServiceLifetime.Singleton, typeof(ODataBatchHandler), s => new MyODataBatchHandler());
                    configureAction.AddService(Microsoft.OData.ServiceLifetime.Singleton, typeof(ODataSerializerProvider), sp => new MyODataSerializerProvider(sp));
                });
```
In previos version I was doing it like this:

```
 app.UseMvc(routeBuilder =>
            {
                routeBuilder.MapODataServiceRoute("Odatatoken", "{contextToken}", configureAction =>
                { }); });
```
When I remove {contextToken} from AddModel it is taking only 176mb. 
I belive something wrong in  `opt.AddModel("{contextToken}", edmModel, configureAction =>`. Memory can grow up, but I don't think I should grow up to 8 gb. Adding dynamic prefix is very important to me. I cannot to migrate to this version if I cannot implement dynamic prefix

**Comment by xuzhg** [comment]
@komdil Interesting.

```C$
opt.AddModel(edmModel, configureAction =>
                {
                    configureAction.AddService(Microsoft.OData.ServiceLifetime.Singleton, typeof(ODataBatchHandler), s => new MyODataBatchHandler());
                    configureAction.AddService(Microsoft.OData.ServiceLifetime.Singleton, typeof(ODataSerializerProvider), sp => new MyODataSerializerProvider(sp));
                });
```
only takes 176mb.

```C#
opt.AddModel("{contextToken}", edmModel,
```
takes 8 gb?

In this case, I have to investigate it to understand it.

**Comment by komdil** [comment]
@xuzhg Yes, in case with {contextToken} it is taking 8gb (even more) memory. It is happening after sending request. Also it is hanging 

![image](https://user-images.githubusercontent.com/50167116/117336900-243a1800-aeb6-11eb-978c-f7683310af7e.png)

**Comment by xuzhg** [comment]
@komdil Thanks for your detail information. I use `builder.UseODataRouteDebug();` and send "$odata" request.

It shows me a very big routing table:

![image](https://user-images.githubusercontent.com/9426627/117361346-c6361200-ae6e-11eb-806e-8f64686a30a5.png)

There are 107 odata endpoints and 84 non-odata endpoints. It could be a big consuming for the memory for the DFA matcher.
I will work with ASP.NET Core team (using your sample) to dig more and improve the state logic in .NET 6.

By the way, there's a dynamic model sample that you can consider to take as a workaround. 
https://github.com/OData/AspNetCoreOData/tree/master/sample/ODataDynamicModel

please take a look and let me know your thoughts.
